contributing/development/handling_compatibility_breakages.rst

```diff
@@ -0,0 +1,168 @@
+.. _doc_handling_compatibility_breakages:
+
+Handling compatibility breakages
+=================================
+
+So you've added a new parameter to a method, changed the return type,
+changed the type of a parameter, or changed its default value,
+and now the automated testing is complaining about compatibility breakages?
+
+Not to worry, there's a system to handle just that!
+
+Compatibility is handled in a few different ways. In GDScript this is simple,
+new arguments are usually given default values when added, so existing scripts
+won't break in such cases.
+
+However in other cases, like GDExtension and C#, the exact way the method is
+declared matters and must be matched exactly. This is handled by creating
+compatibility methods that then get mapped to the new versions of these
+methods. All of this happens *almost* automatically, there are just a few
+things you need to do.
+
+Binding compatibility methods
+-----------------------------
+
+Just like how you use ``ClassDB`` to bind methods, there's a way to bind
+compatibility methods, and conveniently these are named very similarly to their
+ordinary counterparts.
+
+.. code-block:: cpp
+
+    ClassDB::bind_method(D_METHOD("my_method", "arg1", "arg2", "arg3"), &MyClass::my_method, DEFVAL(0));
+    ClassDB::bind_compatibility_method(D_METHOD("my_method", "arg1", "arg2"), &MyClass::_my_method_bind_compat_1234);
+
+The extension validation system
+-------------------------------
+
+You also need to handle updating the expected output of the compatibility checks.
+
+As part of the normal set of checks that run on every PR there's one that checks for compatibility with extensions.
+This checks the API generated by the engine against that of previous versions, as well as files containing the expected
+results.
+
+To tell the system that the changes made in a PR were not by mistake but intentional compatibility breakages you need
+to add the results of this check to the relevant validation file. These files are stored under ``misc/extension_api_validation/``
+and are named in two ways. Either a short name with just one version in it, like ``4.2-stable.expected``, or two versions,
+one after another, like ``4.1-stable_4.2-stable.expected``. Most of the time you only need to care about the first type, which will
+have the version number of the most recently released version.
+
+The specific instructions for what to do when this check fails are at the top of each of these files.
```

Comment:
Might be a bit rambling, any suggestions would be welcome

Reply:
I feel those two generic sections tend to overlap with explanations given below in the practical example, especially the one about binding compat methods which I would expect to explain how to name the compat method, where to place it, etc.

Maybe we could introduce the example from the start and not try to have standalone generic explanations, but use the example throughout to make the generic bits clearer?

Reply:
Dropping the generics, didn't feel any of it was useful to integrate into the practical example